File: cpp/ultra-fast.cpp
#include <iostream>
#include <array>
#include <random>
#include <bitset>
#include <algorithm>
#include <atomic>
#include <thread>

const int ROUNDS = 5;

class UltraFastBaseGame {
private:
    std::atomic<uint16_t> board;
    std::array<uint16_t, 13> subsetCache;
    std::array<int, 2> scores;

    class FastRandom {
    private:
        uint64_t state;
    public:
        explicit FastRandom(uint64_t seed) : state(seed) {}
        
        uint64_t next() {
            state ^= state >> 12;
            state ^= state << 25;
            state ^= state >> 27;
            return state * 0x2545F4914F6CDD1DULL;
        }

        int rollDice() {
            return (next() % 6) + 1 + (next() % 6) + 1;
        }
    };

    void initSubsetCache() {
        for (int i = 2; i <= 12; ++i) {
            subsetCache[i] = (1 << i) - 1;
        }
    }

public:
    UltraFastBaseGame(uint64_t seedA = 0, uint64_t seedB = 0) 
        : board(0xFFF), scores{0, 0} {
        initSubsetCache();
    }

    void play() {
        for (int round = 0; round < ROUNDS; ++round) {
            for (int player = 0; player < 2; ++player) {
                board.store(0xFFF, std::memory_order_relaxed);
                int turnScore = playerTurn(player);
                scores[player] += turnScore;
            }
        }

        std::cout << "Game over. Final scores: Player A = " << scores[0] 
                  << ", Player B = " << scores[1] << std::endl;
        std::cout << "Winner: Player " << (scores[0] < scores[1] ? "A" : "B") << std::endl;
    }

    int playerTurn(int player) {
        FastRandom rng(std::chrono::high_resolution_clock::now().time_since_epoch().count());
        while (true) {
            int roll = rng.rollDice();
            uint16_t mask = 1 << (roll - 1);
            uint16_t subset = subsetCache[roll];

            uint16_t oldBoard, newBoard;
            bool success = false;

            do {
                oldBoard = board.load(std::memory_order_relaxed);
                if (oldBoard & mask) {
                    newBoard = oldBoard & ~mask;
                    success = true;
                } else if ((oldBoard & subset) == subset) {
                    newBoard = oldBoard & ~subset;
                    success = true;
                } else {
                    return __builtin_popcount(oldBoard);
                }
            } while (!board.compare_exchange_weak(oldBoard, newBoard, std::memory_order_release, std::memory_order_relaxed));

            if (newBoard == 0) {
                return 0;
            }
        }
    }
};

File: rust/ultra-fast.rs
// ultra_fast_base_game.rs
use std::sync::atomic::{AtomicU16, Ordering};
use std::time::{SystemTime, UNIX_EPOCH};

const FULL_BOARD: u16 = 0xFFF;

struct FastRandom {
    state: [u64; 2],
}

impl FastRandom {
    fn new(seed1: u64, seed2: u64) -> Self {
        FastRandom { state: [seed1, seed2] }
    }

    fn next(&mut self) -> u64 {
        let s0 = self.state[0];
        let mut s1 = self.state[1];
        let result = s0.wrapping_add(s1);

        s1 ^= s0;
        self.state[0] = s0.rotate_left(55) ^ s1 ^ (s1 << 14);
        self.state[1] = s1.rotate_left(36);

        result
    }

    fn roll_dice(&mut self) -> u16 {
        (self.next() % 11 + 2) as u16
    }
}

pub struct UltraFastBaseGame {
    board: AtomicU16,
    subset_cache: [u16; 13],
    scores: [u32; 2],
}

impl UltraFastBaseGame {
    pub fn new() -> Self {
        let mut game = UltraFastBaseGame {
            board: AtomicU16::new(FULL_BOARD),
            subset_cache: [0; 13],
            scores: [0; 2],
        };
        game.init_subset_cache();
        game
    }

    fn init_subset_cache(&mut self) {
        for i in 2..=12 {
            self.subset_cache[i] = (1 << i) - 1;
        }
    }

    pub fn play(&mut self) {
        let seed = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos() as u64;
        let mut rng = FastRandom::new(seed, seed.wrapping_add(1));

        for _ in 0..5 {
            for player in 0..2 {
                self.board.store(FULL_BOARD, Ordering::Relaxed);
                let turn_score = self.player_turn(player, &mut rng);
                self.scores[player] += turn_score as u32;
            }
        }

        println!("Game over. Final scores: Player A = {}, Player B = {}", 
                 self.scores[0], self.scores[1]);
        println!("Winner: Player {}", if self.scores[0] < self.scores[1] { "A" } else { "B" });
    }

    fn player_turn(&self, _player: usize, rng: &mut FastRandom) -> u16 {
        loop {
            let roll = rng.roll_dice();
            let mask = 1 << (roll - 1);
            let subset = self.subset_cache[roll as usize];

            let result = self.board.fetch_update(Ordering::Release, Ordering::Relaxed, |old_board| {
                if old_board & mask != 0 {
                    Some(old_board & !mask)
                } else if old_board & subset == subset {
                    Some(old_board & !subset)
                } else {
                    None
                }
            });

            match result {
                Ok(new_board) => {
                    if new_board == 0 {
                        return 0;  // Perfect score
                    }
                },
                Err(old_board) => {
                    return old_board.count_ones() as u16;  // Turn ends, return score
                }
            }

            std::hint::spin_loop();
        }
    }
}